### ISSUE TITLE
Bracket.use upon task cancellation reads outdated state if the resource is an MVar

### ISSUE THREAD

**Issue by lukasz-golebiewski** [typelevel/cats-effect#503]
I noticed a weird behavior when using `MVar` together with the `Bracket` typeclass. My initial idea was to use it for synchronization instead of `Ref` together with a `Semaphore` but I hit a corner case which ended with a deadlock. The test below when added to the `BaseMVarJVMSuite` gives an outlook on the problem. I believe the correct behavior would be if the assertion never failed, but it does with:

`org.scalatest.exceptions.TestFailedException: x did not equal -1 the task was cancelled and read an outdated value` 
- where x can range from 0 to 5

In a very unfortunate corner case the fiber which raises the exception will execute as last thus populating the `MVar` with an incorrect value

```scala
test("MVar (concurrent) — bracket.use returns latest state on cancel") {
    def run = {
      import cats._
      import cats.effect.ExitCase._

      def decrement(d: MVar[IO,Int]) = for {
        _ <- Sync[IO].bracketCase(d.take) { v =>
          if(v < 0) MonadError[IO, Throwable].raiseError[Unit](new IllegalArgumentException("Value is below zero")) else d.put(v - 1)
              }{
            case (_, Completed) => IO.unit
            case (previous, Canceled) => IO(assert(previous == -1, s"the task was cancelled and read an outdated value" )) *>
            d.put(previous)
            case (previous, _) => d.put(previous)
          }
        } yield ()

      for {
        d     <- MVar.of[IO, Int](5)
        _    <- Parallel.parSequence(List.fill(100)(decrement(d))).handleError{case e : IllegalArgumentException => List(())}
        res   <- d.read
      } yield res
    }

    for (_ <- 0 until iterations) {
      assert(run.unsafeRunTimed(timeout) == Some(-1), s"; timed-out after $timeout")
    }
  }
```

**Comment by djspiewak** [comment]
This is quite peculiar. What value exactly do you get here?

**Comment by lukasz-golebiewski** [comment]
@djspiewak I wrote it in the description: "x can range from 0 to 5"

**Comment by djspiewak** [comment]
My *guess* as to what's happening here (I'll use `x = 5` as an example, though it would be helpful to know the probability distribution of 0-5 on the results):

1. The very first `d.take` runs
2. `x = 5` and so the conditional is passed
3. `d.put(4)` runs
4. *All* the other valid `decrement`s run
5. `d.take` where `x = -1` runs
6. An error is raised
7. All of the `IO`s are canceled by `parSequence`
8. Step 3 still has not yet *completed* its run, it only completed the `put`
9. The very first `IO` is canceled *at step 3*, despite the fact that it was valid

In other words, you're making the assumption that all valid `decrement`s will *fully* complete (as in, the `bracket` fully resolves) before the first cancelation from the invalid `decrement`s, and this assumption is not necessarily true. It may be something that we can guarantee in situations like this, but I'm not 100% sure of it. It would certainly depend quite a bit on the underlying `IO` interpreter, and so you could probably only rely on the semantic for cats-effect unless other libraries implemented it.

**Comment by djspiewak** [comment]
@alexandru If you have some time, can you weigh in here? I'm frankly not certain exactly what guarantees we should expect from `MVar` in these kinds of situations.

**Comment by djspiewak** [comment]
Since `MVar` is now deprecated in CE2 and doesn't exist at all in CE3, I think we can close this.
